google-cloud-storage/src/main/java/com/google/cloud/storage/WriteFlushStrategy.java

```diff
@@ -51,22 +53,34 @@ private WriteFlushStrategy() {}
    */
   static FlusherFactory fsyncEveryFlush(
       ClientStreamingCallable<WriteObjectRequest, WriteObjectResponse> write) {
-    return (LongConsumer committedTotalBytesCallback,
+    return (String bucketName,
+        LongConsumer committedTotalBytesCallback,
         Consumer<WriteObjectResponse> onSuccessCallback) ->
-        new FsyncEveryFlusher(write, committedTotalBytesCallback, onSuccessCallback);
+        new FsyncEveryFlusher(write, bucketName, committedTotalBytesCallback, onSuccessCallback);
   }
 
   /**
    * Create a {@link Flusher} which will "fsync" only on {@link Flusher#close(WriteObjectRequest)}.
    * Calls to {@link Flusher#flush(List)} will be sent but not synced.
    *
-   * @see FlusherFactory#newFlusher(LongConsumer, Consumer)
+   * @see FlusherFactory#newFlusher(String, LongConsumer, Consumer)
    */
   static FlusherFactory fsyncOnClose(
       ClientStreamingCallable<WriteObjectRequest, WriteObjectResponse> write) {
-    return (LongConsumer committedTotalBytesCallback,
+    return (String bucketName,
+        LongConsumer committedTotalBytesCallback,
         Consumer<WriteObjectResponse> onSuccessCallback) ->
-        new FsyncOnClose(write, committedTotalBytesCallback, onSuccessCallback);
+        new FsyncOnClose(write, bucketName, committedTotalBytesCallback, onSuccessCallback);
+  }
+
+  private static GrpcCallContext contextWithBucketName(String bucketName) {
+    GrpcCallContext ret = GrpcCallContext.createDefault();
+    if (bucketName != null && !bucketName.isEmpty()) {
+      return ret.withExtraHeaders(
+          ImmutableMap.of(
+              "x-goog-request-params", ImmutableList.of(String.format("bucket=%s", bucketName))));
```

Comment:
oh, i thought GAPIC would do this for us?

Reply:
Unfortunately it can't do it for us on the ClientStream, it's technically possible to open a client stream without a request at open time. Due to this, we need to pass it along when we first create the stream.